### PR TITLE
C++ autograd tutorial

### PR DESCRIPTION
**Pull request by yf225**
Accompanying example code: https://github.com/pytorch/examples/pull/745.

### CODE REVIEW

File: advanced_source/cpp_autograd.rst [hunk 1]
```diff
@@ -0,0 +1,442 @@
...
+  auto x = torch::ones({2, 2}, torch::requires_grad());
+  std::cout << x << std::endl;
+
+Out:
+
+.. code-block:: shell
+
+  1 1
+  1 1
+  [ CPUFloatType{2,2} ]
```
Comment (albanD) [inline advanced_source/cpp_autograd.rst:31]: The cpp Tensor print does not show requires_grad and the grad_fn like we do in python?

File: advanced_source/cpp_autograd.rst [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+Out:
+
+.. code-block:: shell
+
+    102.4000
+   1024.0000
+      0.1024
+  [ CPUFloatType{3} ]
+
+You can also stop autograd from tracking history on Tensors with ``.requires_grad() == true``
```
Comment (albanD) [inline advanced_source/cpp_autograd.rst:174]: I don't think this reads as well as the python version.
I think you can simply state "on Tensors that require gradients".

File: advanced_source/cpp_autograd.rst [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+  
+  // Calculate loss
+  auto target = torch::randn({3, 3});
+  auto loss = torch::nn::MSELoss()(output, target);
+  
+  // Use norm of gradients as penalty
+  auto grad_output = torch::ones_like(output);
+  auto gradient = torch::autograd::grad({output}, {input}, /*grad_outputs=*/{grad_output},
+                  /*retain_graph=*/true, /*create_graph=*/true,
+                  /*allow_unused=*/true)[0];
```
Comment (albanD) [inline advanced_source/cpp_autograd.rst:241]: Why do you set allow_unused here?
Also retain_graph is already set to true if not specified and create_graph=true.

File: advanced_source/cpp_autograd.rst [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+  gradient = gradient.view({-1, 1});
```
Comment (albanD) [inline advanced_source/cpp_autograd.rst:242]: What is the goal of this view? Especially if you take the norm on dim 1 below.

File: advanced_source/cpp_autograd.rst [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+and ``torch::autograd::grad``
+(`link <https://pytorch.org/cppdocs/api/function_namespacetorch_1_1autograd_1a1e03c42b14b40c306f9eb947ef842d9c.html>`_)
+for more information on how to use them.
+
+Using custom autograd function in C++
+-------------------------------------
+
+(Adapted from `this tutorial <https://pytorch.org/docs/stable/notes/extending.html#extending-torch-autograd>`_)
+
+Adding operations to ``torch::autograd`` requires implementing a new ``torch::autograd::Function``
```
Comment (albanD) [inline advanced_source/cpp_autograd.rst:270]: "operations" here might not be the best naming (even though it is in the other tutorial).
Maybe "Adding a new elementary operation to"

File: advanced_source/cpp_autograd.rst [hunk 6]
```diff
@@ -0,0 +1,442 @@
...
+  
+  using namespace torch::autograd;
+  
+  // Inherit from Function
+  class LinearFunction : public Function<LinearFunction> {
+   public:
+    // Note that both forward and backward are static functions
+  
+    // bias is an optional argument
+    static Variable forward(AutogradContext *ctx, Variable input, Variable weight, Variable bias = Variable()) {
```
Comment (albanD) [inline advanced_source/cpp_autograd.rst:291]: nit: The default value for bias should be a Tensor not a Variable right?

File: advanced_source/cpp_autograd.rst [hunk 7]
```diff
@@ -0,0 +1,442 @@
...
+      ctx->save_for_backward({input, weight, bias});
+      auto output = input.mm(weight.t());
+      if (bias.defined()) {
+        output += bias.unsqueeze(0).expand_as(output);
+      }
+      return output;
+    }
+  
+    static variable_list backward(AutogradContext *ctx, variable_list grad_outputs) {
+      auto saved = ctx->get_saved_variables();
```
Comment (albanD) [inline advanced_source/cpp_autograd.rst:301]: Why is this not `get_saved_tensors()` ?

File: advanced_source/cpp_autograd.rst [hunk 8]
```diff
@@ -0,0 +1,442 @@
...
+    static Variable forward(AutogradContext *ctx, Variable variable, double constant) {
+      // ctx is a context object that can be used to stash information
+      // for backward computation
+      ctx->saved_data["constant"] = constant;
+      return variable * constant;
+    }
+  
+    static variable_list backward(AutogradContext *ctx, variable_list grad_outputs) {
+      // We return as many input gradients as there were arguments.
+      // Gradients of non-Tensor arguments to forward must be `Variable()`.
```
Comment (albanD) [inline advanced_source/cpp_autograd.rst:362]: Here again, shouldn't these be Tensors?

File: advanced_source/cpp_autograd.rst [hunk 9]
```diff
@@ -0,0 +1,442 @@
...
+| ``torch.Tensor.is_leaf``       | ``torch::Tensor::is_leaf`` (`link <https://pytorch.org/cppdocs/api/classat_1_1_tensor.html#_CPPv4NK2at6Tensor7is_leafEv>`_)                                            |
++--------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+| ``torch.Tensor._version``      | ``torch::Tensor::_version`` (`link <https://pytorch.org/cppdocs/api/classat_1_1_tensor.html#_CPPv4NK2at6Tensor8_versionEv>`_)                                          |
++--------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+| ``torch.Tensor._base``         | ``torch::Tensor::_base`` (`link <https://pytorch.org/cppdocs/api/classat_1_1_tensor.html#_CPPv4NK2at6Tensor5_baseEv>`_)                                                |
```
Comment (albanD) [inline advanced_source/cpp_autograd.rst:428]: Do you really want to mention the internal methods here?
If so, you're missing _is_view().

File: advanced_source/cpp_autograd.rst [hunk 10]
```diff
@@ -0,0 +1,442 @@
...
+
+After translation, most of your Python autograd code should just work in C++.
+If that's not the case, please file a bug report at `GitHub issues <https://github.com/pytorch/pytorch/issues>`_
+and we will fix it as soon as possible.
+
+Conclusion
+----------
+
+You should now have a good overview of PyTorch's C++ autograd API.
+You can find the code examples displayed in this note (yf225 TODO update link) `here
```
Comment (albanD) [inline advanced_source/cpp_autograd.rst:439]: nit: TODO

File: advanced_source/cpp_autograd.rst (hunk 1 above)
Comment (yf225) [reply advanced_source/cpp_autograd.rst:31]: No it doesn't :( We might want to fix it though to be consistent with Python tensor print. I filed an issue here: https://github.com/pytorch/pytorch/issues/36004.

File: advanced_source/cpp_autograd.rst (hunk 7 above)
Comment (yf225) [reply advanced_source/cpp_autograd.rst:299]: Yeah I think this should be `get_saved_tensors()`, and my guess is that this API was created before the Variable/Tensor merge was completed, and the author wanted to make a distinction between tensors and variables. Now that we don't have this distinction anymore, I think we should change it to match the Python API: https://github.com/pytorch/pytorch/issues/36099.

Comment (yf225) [reply advanced_source/cpp_autograd.rst:299]: I renamed all `Variable`s to `Tensor` now, to minimize the inconsistency with Python to only this `get_saved_variables()` function :)

Comment (albanD) [reply advanced_source/cpp_autograd.rst:299]: Yes I guess it is just a remain of the cpp Tensor/Variable merge.

**Comment by yf225** [comment]
Note to self: merge https://github.com/pytorch/examples/pull/745 first and update link in this PR
